Autopay Customer Invoices

nospame opened the pull request:
## Product Description
Allows Customer Invoices (for "customer billing" accounts) to be autopaid. There's no user-facing (customer or accounting admin) indication this wouldn't work, so likely it was just never implemented.

## Technical Summary
https://dimagi.atlassian.net/browse/SAAS-18346
Follows the existing implementation for autopaying single-project Invoices but adds a new method `pay_autopayable_customer_invoices`, called by the `pay_autopay_invoices` task to explicitly pay Customer Invoices. Updates email context, templates, and logging to indicate "invoice"/"project" or "customer invoice"/"billing account" as needed. Also adds translation tags in email templates where they were not previously.

Review by commit.

## Safety Assurance

### Safety story
Tested on staging by setting up a customer billing account, generating customer invoices, and calling `pay_autopayable_customer_invoices` with the account name specified, checking that the charge was made through Stripe's testing API, the invoice was credited correctly and marked as paid, and the email was sent out as expected. Did the same with regular invoices and `pay_autopayable_invoices` to confirm that functionality is unaffected as well.

Automated tests are also added or updated here for `CustomerInvoice`s to mirror what what we do for the `Invoice` autopay process.

### Automated test coverage
- `TestBillingAutoPayCustomerInvoices` is added mirroring the renamed `TestBillingAutoPayInvoices` (previously just `TestBillingAutoPay`.
- Autopay email context tests `TestGetContextToSendAutopayFailedEmail` and `TestGetContextToSendPurchaseReceipt` are also updated to ensure expected context with the changes made to accommodate customer invoices.

### QA Plan
Not planning it.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Allow _pay_invoice to autopay CustomerInvoices

nospame committed:
Update email context and templates for CustomerInvoices

nospame committed:
Update TestGetContext classes

nospame committed:
Fix bug in make_payment_towards_invoice

Would not have worked with customer invoices before, even though it could be called through CustomerInvoice.pay_invoice

nospame committed:
Pay autopay customer invoices

nospame committed:
Add TestBillingAutoPayCustomerInvoices

nospame committed:
Add API-based TestBillingAutoPayCustomerInvoices

biyeun commented:
note for testing on staging: would be good to ensure non customer invoices are still working as expected, since it also touches the same code

biyeun commented on corehq/apps/accounting/task_utils.py:
```diff
@@ -17,27 +19,37 @@
-        'billing_info_url': absolute_reverse('domain_update_billing_info', args=[domain]),
+        'is_customer_invoice': is_customer_invoice,
+        'subscription_plan': subscription_plan,
+        'domain_url': absolute_reverse('dashboard_default', args=[domain_name]),
```
might be clearer to leave these out if `domain_name` is `None`

nospame replied:
https://github.com/dimagi/commcare-hq/pull/37083/commits/81cab194e92d9869e8fa55570ee940a15f4c7bf6

biyeun commented on corehq/apps/accounting/tasks.py:
```diff
@@ -519,8 +519,12 @@ def create_wire_credits_invoice(domain_name,
-def send_purchase_receipt(payment_record_id, domain, template_html, template_plaintext, additional_context):
-    context = get_context_to_send_purchase_receipt(payment_record_id, domain, additional_context)
+def send_purchase_receipt(
+    payment_record_id, domain_name, account_name, template_html, template_plaintext, additional_context
```
pass in account id and lookup based on id

biyeun commented on corehq/apps/accounting/templates/accounting/email/invoice_receipt.txt:
```diff
@@ -1,23 +1,41 @@
+Transaction No.: {{ transaction_id }}
+{% endblocktrans %}
+{% else %}
+{% blocktrans %}
```
check spacing between paragraphs in text version

biyeun commented on corehq/apps/accounting/payment_handlers.py:
```diff
@@ -383,6 +383,21 @@ def pay_autopayable_invoices(self, date_due=Ellipsis, domain=None):
             except Exception as e:
                 log_accounting_error("Error autopaying invoice %d: %s" % (invoice.id, e))
 
+    def pay_autopayable_customer_invoices(self, date_due=Ellipsis, account=None):
```
specify `account_name` vs `account`?

biyeun commented on corehq/apps/accounting/tests/test_autopay.py:
```diff
@@ -172,6 +183,97 @@ def _run_autopay(self):
+
+        self.original_outbox_length = len(mail.outbox)
+        self._run_autopay()
+        self._assert_no_side_effects()
```
could we check that the balance of the invoice is not affected

nospame committed:
Look up BillingAccount based on id, not name

Adds a related test case to see that account name is used when there is no domain_name

nospame committed:
Update get_context_to_send_autopay_failed_email

- Comment on get_domains() for CustomerInvoices
- Don't include useless domain_name urls

nospame committed:
Fix typo

nospame committed:
Specify account_name as kwarg in pay_autopayable_customer_invoices

nospame committed:
Check invoice total is unchanged when stripe fails

nospame committed:
Remove whitespace in autopay email txt templates

Because these are whitespace-sensitive, newlines and indendation can make output look funky.

nospame commented on corehq/apps/accounting/task_utils.py:
```diff
@@ -17,27 +19,37 @@
-    autopay_card = payment_method.get_autopay_card(subscription.account)
-    domain = invoice.get_domain()
+    autopay_card = payment_method.get_autopay_card(account)
+    domain_name = invoice.get_domain()
```
https://github.com/dimagi/commcare-hq/pull/37083/commits/81cab194e92d9869e8fa55570ee940a15f4c7bf6
